### PR TITLE
Add make update to credentials

### PR DESCRIPTION
**Pull request by jansenk**
@edx/masters-neem
EDUCATOR-4369

branch building off of #657

add the following make targets:
make upgrade - recompile and upgrade requirements *.in to *.txt with pip-compile
make piptools - install pip-tools

changed make requirements(-production) to use pip-sync on their respective files rather than pip install

unpinned requirements that could be unpinned, also moved around some requirements. we were previously only installing (not removing) dependencies (pip install vs pip-sync)

---

Make sure that the following steps are done before rebasing/merging

  - [ ] Request a review if desired
  - [ ] Squash/Fixup your branch to one commit
  - Config/Dependency Changes (e.g. config files, scripts that are used for provisioning etc.)
    - [ ] Make sure you have updated [edx/configuration](https://github.com/edx/configuration) and [edx/devstack](https://github.com/edx/devstack) if necessary
    - [ ] Make sure the change builds successfully in a sandbox
  - UI Changes 
    - [ ] Consider other browsers (e.g. Firefox)
    - [ ] Check mobile view
    - [ ] Consider accessibility (e.g. Run aXe on the page)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -135,3 +140,16 @@ validate_translations: ## Test translations files
 	cd credentials && i18n_tool validate -v --check-all
 
 check_translations_up_to_date: fake_translations detect_changed_source_translations ## Install fake translations and check if translation files are up-to-date
+
+piptools: 
+	pip install -q -r requirements/pip_tools.txt
+
+export CUSTOM_COMPILE_COMMAND = make upgrade
+upgrade: piptools ## update the requirements/*.txt files with the latest packages satisfying requirements/*.in
+	pip-compile --rebuild --upgrade -o requirements/pip_tools.txt requirements/pip_tools.in
```
Comment (kdmccormick) [inline Makefile]: `--rebuild` clears pip-compile's caches, which I think is good for the first call, but afterwards we can probably leave the cache intact to speed up the build.

File: Makefile [hunk 2]
```diff
@@ -77,7 +82,7 @@ exec-clean: ## Remove all generated files from a container
 	docker exec -t credentials bash -c 'source /edx/app/credentials/credentials_env && cd /edx/app/credentials/credentials/ && make clean'
 
 exec-requirements:
-	docker exec -t credentials bash -c 'source /edx/app/credentials/credentials_env && cd /edx/app/credentials/credentials/ && make requirements && make production-requirements'
+	docker exec -t credentials bash -c 'source /edx/app/credentials/credentials_env && cd /edx/app/credentials/credentials/ && make all-requirements'
```
Comment (kdmccormick) [inline Makefile]: Do you know why we'd want `exec-requirements` to install all requirements?

File: Makefile [hunk 3]
```diff
@@ -1,7 +1,7 @@
 .DEFAULT_GOAL := tests
 NODE_BIN=./node_modules/.bin
 
-.PHONY: requirements
+.PHONY: requirements upgrade piptools
```
Comment (kdmccormick) [inline Makefile]: Thanks for remembering to do update this haha. Throw `production-requirements` (and anything else that doesn't generate a file with the same name as the recipe) in there too.

File: requirements/base.in [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+#   * verify that the dependency has a license compatible with AGPLv3
+#   * confirm that it has no system requirements beyond what we already install
+#   * run "make upgrade" to update the detailed requirements files
+
+-c constraints.txt
+
+analytics-python
+coreapi
+django<2.0
+django-debug-toolbar==1.6
```
Comment (kdmccormick) [inline requirements/base.in]: Any remaining pins can go in `constraints.txt`.

File: requirements/production.in [hunk 5]
```diff
@@ -0,0 +1,18 @@
+# Dependencies for running the credentials IDA in a production environement
+#
+# DON'T JUST ADD NEW DEPENDENCIES!!!
+#
+# If you open a pull request that adds a new dependency, you should:
+#   * verify that the dependency has a license compatible with AGPLv3
+#   * confirm that it has no system requirements beyond what we already install
+#   * run "make upgrade" to update the detailed requirements files
+
+-r base.in
```
Comment (kdmccormick) [inline requirements/production.in]: So this is actually something I screwed up in Registrar and didn't realize was wrong until recently -- we actually want to include `-r base.txt`, not `-r base.in` (same for the rest of the `-r`s in this PR).

The reasoning is: `base.txt` specifies a bunch of pinned requirements, and we want those to pins to stay the same when we expand it into `test.txt`, `production.txt`, etc. If we include the raw `base.in`, though, then the core requirements in test, production, dev, etc. may be pinned at different versions.

Comment (jansenk) [reply requirements/production.in]: Oh, alright, I actually made this change because that's what we have in registrar.

File: Makefile (hunk 2 above)
Comment (jansenk) [reply Makefile]: No, I don't really understand why we want that

File: Makefile (hunk 1 above)
Comment (iloveagent57) [reply Makefile]: FWIW, one of our other repos does include the `--rebuild` flag (edx-platform), seems like most other IDAs do not.  Probably not a big deal either way.

File: requirements/production.in (hunk 5 above)
Comment (kdmccormick) [reply requirements/production.in]: That's totally fair and we should probably change Registrar at some point :P

File: Makefile (hunk 2 above)
Comment (kdmccormick) [reply Makefile]: It looks like we use it on Travis: https://github.com/edx/credentials/blob/master/.travis.yml#L33-L37

May as well leave it as-is for now. I like the approach you took to make it have all requirements.

File: requirements/base.in [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+pillow
+pygments
+python-memcached
+pytz
+requests
+six
+xss-utils
+
+# TODO Install in configuration
+git+https://github.com/edx/credentials-themes.git@0.1.28#egg=edx_credentials_themes==0.1.28
```
Comment (kdmccormick) [inline requirements/base.in]: This version doesn't actually exist, so pip-compile just tries a few times and complains and then chooses 0.1.29. Might as well just specify 0.1.29.

File: Makefile [hunk 7]
```diff
@@ -135,3 +140,16 @@ validate_translations: ## Test translations files
...
 
 check_translations_up_to_date: fake_translations detect_changed_source_translations ## Install fake translations and check if translation files are up-to-date
+
+piptools: 
+	pip install -q -r requirements/pip_tools.txt
+
+export CUSTOM_COMPILE_COMMAND = make upgrade
+upgrade: piptools ## update the requirements/*.txt files with the latest packages satisfying requirements/*.in
+	pip-compile --rebuild --upgrade -o requirements/pip_tools.txt requirements/pip_tools.in
+	pip-compile --rebuild --upgrade -o requirements/base.txt requirements/base.in
```
Comment (kdmccormick) [inline Makefile]: Gonna advocate for only `--rebuild`ing on the first call again. Rebuilding every time about doubles the `make upgrade` time for me.

File: requirements/base.in [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+pillow
+pygments
+python-memcached
+pytz
+requests
+six
+xss-utils
+
+# TODO Install in configuration
+git+https://github.com/edx/credentials-themes.git@0.1.28#egg=edx_credentials_themes==0.1.27
```
Comment (kdmccormick) [inline requirements/base.in]: Can you change `credentials-themes.git@0.1.28` to `credentials-themes.git@0.1.27` as well?

(What happened is that whoever released `credentials-themes 0.1.28` forgot to bump the version number in the code, so the release named `0.1.28` is still actually `0.1.27`)

**Pull request merged**